支持Doris的读写插件

Doris是兼容MySQL的，独写使用MySQL插件都可以，但是写入的时候使用insert into的方式，Doris会使用FE节点解析sql，导致速度比较慢，所以有没有什么办法开发独立的，速度比较快的Doris读写插件，看官方文档只说 insert into 不适合用在生产环境，转而使用 insert into select 方式，现在这种方式不适合 DataX，但是它又不像PostgresSQL那样提供了类似 COPY FROM 这样的高效数据插入方式，推荐使用Stream Load的方式导入，希望作者可以内部转换成Stream Load，以此来提高导入速度！

官方示例代码(<https://github.com/apache/incubator-doris/blob/master/samples/stream_load/java/DorisStreamLoad.java>) 

```java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import org.apache.commons.codec.binary.Base64;
import org.apache.http.HttpHeaders;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPut;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.DefaultRedirectStrategy;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
/**
 * This class is a java demo for doris stream load
 *
 * The pom.xml dependency:
 *
 *         <dependency>
 *             <groupId>org.apache.httpcomponents</groupId>
 *             <artifactId>httpclient</artifactId>
 *             <version>4.5.3</version>
 *         </dependency>
 *
 * How to use:
 *
 * 1 create a table in doris with any mysql client
 *
 * CREATE TABLE `stream_test` (
 *   `id` bigint(20) COMMENT "",
 *   `id2` bigint(20) COMMENT "",
 *   `username` varchar(32) COMMENT ""
 * ) ENGINE=OLAP
 * DUPLICATE KEY(`id`)
 * DISTRIBUTED BY HASH(`id`) BUCKETS 20;
 *
 *
 * 2 change the Doris cluster, db, user config in this class
 *
 * 3 run this class, you should see the following output:
 *
 * {
 *     "TxnId": 27,
 *     "Label": "39c25a5c-7000-496e-a98e-348a264c81de",
 *     "Status": "Success",
 *     "Message": "OK",
 *     "NumberTotalRows": 10,
 *     "NumberLoadedRows": 10,
 *     "NumberFilteredRows": 0,
 *     "NumberUnselectedRows": 0,
 *     "LoadBytes": 50,
 *     "LoadTimeMs": 151
 * }
 *
 * Attention:
 *
 * 1 wrong dependency version(such as 4.4) of httpclient may cause shaded.org.apache.http.ProtocolException
 *   Caused by: shaded.org.apache.http.ProtocolException: Content-Length header already present
 *     at shaded.org.apache.http.protocol.RequestContent.process(RequestContent.java:96)
 *     at shaded.org.apache.http.protocol.ImmutableHttpProcessor.process(ImmutableHttpProcessor.java:132)
 *     at shaded.org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:182)
 *     at shaded.org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
 *     at shaded.org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
 *     at shaded.org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
 *
 *2 run this class more than once, the status code for http response is still ok, and you will see
 *  the following output:
 *
 * {
 *     "TxnId": -1,
 *     "Label": "39c25a5c-7000-496e-a98e-348a264c81de",
 *     "Status": "Label Already Exists",
 *     "ExistingJobStatus": "FINISHED",
 *     "Message": "Label [39c25a5c-7000-496e-a98e-348a264c81de"] has already been used.",
 *     "NumberTotalRows": 0,
 *     "NumberLoadedRows": 0,
 *     "NumberFilteredRows": 0,
 *     "NumberUnselectedRows": 0,
 *     "LoadBytes": 0,
 *     "LoadTimeMs": 0
 * }
 * 3 when the response statusCode is 200, that doesn't mean your stream load is ok, there may be still
 *   some stream problem unless you see the output with 'ok' message
 */
public class DorisStreamLoad {
    private final static String DORIS_HOST = "xxx.com";
    private final static String DORIS_DB = "test";
    private final static String DORIS_TABLE = "stream_test";
    private final static String DORIS_USER = "root";
    private final static String DORIS_PASSWORD = "xxx";
    private final static int DORIS_HTTP_PORT = 8410;

    private void sendData(String content) throws Exception {
        final String loadUrl = String.format("http://%s:%s/api/%s/%s/_stream_load",
                DORIS_HOST,
                DORIS_HTTP_PORT,
                DORIS_DB,
                DORIS_TABLE);

        final HttpClientBuilder httpClientBuilder = HttpClients
                .custom()
                .setRedirectStrategy(new DefaultRedirectStrategy() {
                    @Override
                    protected boolean isRedirectable(String method) {
                        return true;
                    }
                });

        try (CloseableHttpClient client = httpClientBuilder.build()) {
            HttpPut put = new HttpPut(loadUrl);
            StringEntity entity = new StringEntity(content, "UTF-8");
            put.setHeader(HttpHeaders.EXPECT, "100-continue");
            put.setHeader(HttpHeaders.AUTHORIZATION, basicAuthHeader(DORIS_USER, DORIS_PASSWORD));
            // the label header is optional, not necessary
            // use label header can ensure at most once semantics
            put.setHeader("label", "39c25a5c-7000-496e-a98e-348a264c81de");
            put.setEntity(entity);

            try (CloseableHttpResponse response = client.execute(put)) {
                String loadResult = "";
                if (response.getEntity() != null) {
                    loadResult = EntityUtils.toString(response.getEntity());
                }
                final int statusCode = response.getStatusLine().getStatusCode();
                // statusCode 200 just indicates that doris be service is ok, not stream load
                // you should see the output content to find whether stream load is success
                if (statusCode != 200) {
                    throw new IOException(
                            String.format("Stream load failed, statusCode=%s load result=%s", statusCode, loadResult));
                }

                System.out.println(loadResult);
            }
        }
    }

    private String basicAuthHeader(String username, String password) {
        final String tobeEncode = username + ":" + password;
        byte[] encoded = Base64.encodeBase64(tobeEncode.getBytes(StandardCharsets.UTF_8));
        return "Basic " + new String(encoded);
    }

    public static void main(String[] args) throws Exception {
        int id1 = 1;
        int id2 = 10;
        String id3 = "张三";
        int rowNumber = 10;
        String oneRow = id1 + "\t" + id2 + "\t" + id3 + "\n";

        StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < rowNumber; i++) {
            stringBuilder.append(oneRow);
        }

        //in doris 0.9 version, you need to comment this line
        //refer to https://github.com/apache/incubator-doris/issues/783
        stringBuilder.deleteCharAt(stringBuilder.length() - 1);

        String loadData = stringBuilder.toString();
        DorisStreamLoad dorisStreamLoad = new DorisStreamLoad();
        dorisStreamLoad.sendData(loadData);
    }
}
```

> 官方示例代码(https://github.com/apache/incubator-doris/blob/master/samples/stream_load/java/DorisStreamLoad.java)
> 
> ```java
> // Licensed to the Apache Software Foundation (ASF) under one
> // or more contributor license agreements.  See the NOTICE file
> // distributed with this work for additional information
> // regarding copyright ownership.  The ASF licenses this file
> // to you under the Apache License, Version 2.0 (the
> // "License"); you may not use this file except in compliance
> // with the License.  You may obtain a copy of the License at
> //
> //   http://www.apache.org/licenses/LICENSE-2.0
> //
> // Unless required by applicable law or agreed to in writing,
> // software distributed under the License is distributed on an
> // "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
> // KIND, either express or implied.  See the License for the
> // specific language governing permissions and limitations
> // under the License.
> 
> import org.apache.commons.codec.binary.Base64;
> import org.apache.http.HttpHeaders;
> import org.apache.http.client.methods.CloseableHttpResponse;
> import org.apache.http.client.methods.HttpPut;
> import org.apache.http.entity.StringEntity;
> import org.apache.http.impl.client.CloseableHttpClient;
> import org.apache.http.impl.client.DefaultRedirectStrategy;
> import org.apache.http.impl.client.HttpClientBuilder;
> import org.apache.http.impl.client.HttpClients;
> import org.apache.http.util.EntityUtils;
> 
> import java.io.IOException;
> import java.nio.charset.StandardCharsets;
> /**
>  * This class is a java demo for doris stream load
>  *
>  * The pom.xml dependency:
>  *
>  *         <dependency>
>  *             <groupId>org.apache.httpcomponents</groupId>
>  *             <artifactId>httpclient</artifactId>
>  *             <version>4.5.3</version>
>  *         </dependency>
>  *
>  * How to use:
>  *
>  * 1 create a table in doris with any mysql client
>  *
>  * CREATE TABLE `stream_test` (
>  *   `id` bigint(20) COMMENT "",
>  *   `id2` bigint(20) COMMENT "",
>  *   `username` varchar(32) COMMENT ""
>  * ) ENGINE=OLAP
>  * DUPLICATE KEY(`id`)
>  * DISTRIBUTED BY HASH(`id`) BUCKETS 20;
>  *
>  *
>  * 2 change the Doris cluster, db, user config in this class
>  *
>  * 3 run this class, you should see the following output:
>  *
>  * {
>  *     "TxnId": 27,
>  *     "Label": "39c25a5c-7000-496e-a98e-348a264c81de",
>  *     "Status": "Success",
>  *     "Message": "OK",
>  *     "NumberTotalRows": 10,
>  *     "NumberLoadedRows": 10,
>  *     "NumberFilteredRows": 0,
>  *     "NumberUnselectedRows": 0,
>  *     "LoadBytes": 50,
>  *     "LoadTimeMs": 151
>  * }
>  *
>  * Attention:
>  *
>  * 1 wrong dependency version(such as 4.4) of httpclient may cause shaded.org.apache.http.ProtocolException
>  *   Caused by: shaded.org.apache.http.ProtocolException: Content-Length header already present
>  *     at shaded.org.apache.http.protocol.RequestContent.process(RequestContent.java:96)
>  *     at shaded.org.apache.http.protocol.ImmutableHttpProcessor.process(ImmutableHttpProcessor.java:132)
>  *     at shaded.org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:182)
>  *     at shaded.org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
>  *     at shaded.org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
>  *     at shaded.org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
>  *
>  *2 run this class more than once, the status code for http response is still ok, and you will see
>  *  the following output:
>  *
>  * {
>  *     "TxnId": -1,
>  *     "Label": "39c25a5c-7000-496e-a98e-348a264c81de",
>  *     "Status": "Label Already Exists",
>  *     "ExistingJobStatus": "FINISHED",
>  *     "Message": "Label [39c25a5c-7000-496e-a98e-348a264c81de"] has already been used.",
>  *     "NumberTotalRows": 0,
>  *     "NumberLoadedRows": 0,
>  *     "NumberFilteredRows": 0,
>  *     "NumberUnselectedRows": 0,
>  *     "LoadBytes": 0,
>  *     "LoadTimeMs": 0
>  * }
>  * 3 when the response statusCode is 200, that doesn't mean your stream load is ok, there may be still
>  *   some stream problem unless you see the output with 'ok' message
>  */
> public class DorisStreamLoad {
>     private final static String DORIS_HOST = "xxx.com";
>     private final static String DORIS_DB = "test";
>     private final static String DORIS_TABLE = "stream_test";
>     private final static String DORIS_USER = "root";
>     private final static String DORIS_PASSWORD = "xxx";
>     private final static int DORIS_HTTP_PORT = 8410;
> 
>     private void sendData(String content) throws Exception {
>         final String loadUrl = String.format("http://%s:%s/api/%s/%s/_stream_load",
>                 DORIS_HOST,
>                 DORIS_HTTP_PORT,
>                 DORIS_DB,
>                 DORIS_TABLE);
> 
>         final HttpClientBuilder httpClientBuilder = HttpClients
>                 .custom()
>                 .setRedirectStrategy(new DefaultRedirectStrategy() {
>                     @Override
>                     protected boolean isRedirectable(String method) {
>                         return true;
>                     }
>                 });
> 
>         try (CloseableHttpClient client = httpClientBuilder.build()) {
>             HttpPut put = new HttpPut(loadUrl);
>             StringEntity entity = new StringEntity(content, "UTF-8");
>             put.setHeader(HttpHeaders.EXPECT, "100-continue");
>             put.setHeader(HttpHeaders.AUTHORIZATION, basicAuthHeader(DORIS_USER, DORIS_PASSWORD));
>             // the label header is optional, not necessary
>             // use label header can ensure at most once semantics
>             put.setHeader("label", "39c25a5c-7000-496e-a98e-348a264c81de");
>             put.setEntity(entity);
> 
>             try (CloseableHttpResponse response = client.execute(put)) {
>                 String loadResult = "";
>                 if (response.getEntity() != null) {
>                     loadResult = EntityUtils.toString(response.getEntity());
>                 }
>                 final int statusCode = response.getStatusLine().getStatusCode();
>                 // statusCode 200 just indicates that doris be service is ok, not stream load
>                 // you should see the output content to find whether stream load is success
>                 if (statusCode != 200) {
>                     throw new IOException(
>                             String.format("Stream load failed, statusCode=%s load result=%s", statusCode, loadResult));
>                 }
> 
>                 System.out.println(loadResult);
>             }
>         }
>     }
> 
>     private String basicAuthHeader(String username, String password) {
>         final String tobeEncode = username + ":" + password;
>         byte[] encoded = Base64.encodeBase64(tobeEncode.getBytes(StandardCharsets.UTF_8));
>         return "Basic " + new String(encoded);
>     }
> 
>     public static void main(String[] args) throws Exception {
>         int id1 = 1;
>         int id2 = 10;
>         String id3 = "张三";
>         int rowNumber = 10;
>         String oneRow = id1 + "\t" + id2 + "\t" + id3 + "\n";
> 
>         StringBuilder stringBuilder = new StringBuilder();
>         for (int i = 0; i < rowNumber; i++) {
>             stringBuilder.append(oneRow);
>         }
> 
>         //in doris 0.9 version, you need to comment this line
>         //refer to https://github.com/apache/incubator-doris/issues/783
>         stringBuilder.deleteCharAt(stringBuilder.length() - 1);
> 
>         String loadData = stringBuilder.toString();
>         DorisStreamLoad dorisStreamLoad = new DorisStreamLoad();
>         dorisStreamLoad.sendData(loadData);
>     }
> }
> ```

如果放在插件里应该如何处理

--- a/docs/src/main/sphinx/writer/doriswriter.md
+++ b/docs/src/main/sphinx/writer/doriswriter.md
@@ -1,0 +1,184 @@
+# DorisWriter
+
+DorisWriter 插件实现了以流式方式加载数据到 [Doris](http://doris.incubator.apache.org/master/zh-CN/) ，其实现上是通过访问 Doris http 连接(8030)，然后通过 [stream load](http://doris.incubator.apache.org/master/zh-CN/administrator-guide/load-data/stream-load-manual.html)
+加载数据到数据中，相比 `insert into` 方式效率要高不少，也是官方推荐的生产环境下的数据加载方式。
+
+Doris 是一个兼容 MySQL 协议的数据库后端，因此读取 Doris 可以使用 [MySQLReader](../reader/mysqlreader.html) 进行访问。 
+
+## 示例
+
+假定要写入的表的建表语句如下：
+
+```sql
+CREATE DATABASE example_db;
+CREATE TABLE example_db.table1
+(
+    siteid INT DEFAULT '10',
+    citycode SMALLINT,
+    username VARCHAR(32) DEFAULT '',
+    pv BIGINT SUM DEFAULT '0'
+)
+AGGREGATE KEY(siteid, citycode, username)
+DISTRIBUTED BY HASH(siteid) BUCKETS 10
+PROPERTIES("replication_num" = "1");
+```
+
+下面配置一个从内存读取数据，然后写入到 doris 表的配置文件
+
+```json
+{
+  "job": {
+    "setting": {
+      "speed": {
+        "channel": 2
+      }
+    },
+    "content": [
+      {
+        "writer": {
+          "name": "doriswriter",
+          "parameter": {
+            "username": "test",
+            "password": "123456",
+            "batchSize": 1024,
+            "connection": [
+              {
+                "table": "table1",
+                "database": "example_db",
+                "endpoint": "http://127.0.0.1:8030/"
+              }
+            ]
+          }
+        },
+        "reader": {
+          "name": "streamreader",
+          "parameter": {
+            "column": [
+              {
+                "random": "1,500",
+                "type": "long"
+              },
+              {
+                "random": "1,127",
+                "type": "long"
+              },
+              {
+                "value": "this is a text",
+                "type": "string"
+              },
+              {
+                "random": "5,200",
+                "type": "long"
+              }
+            ],
+            "sliceRecordCount": 100
+          }
+        }
+      }
+    ]
+  }
+}
+```
+
+将上述配置文件保存为 `job/stream2doris.json`
+
+执行下面的命令
+
+```shell
+bin/datax.py job/stream2doris.json
+```
+
+输出类似如下：
+
+```
+2021-02-23 15:22:57.851 [main] INFO  VMInfo - VMInfo# operatingSystem class => sun.management.OperatingSystemImpl
+2021-02-23 15:22:57.871 [main] INFO  Engine -
+{
+"content":[
+{
+"reader":{
+    "parameter":{
+            "column":[
+                    {
+                            "random":"1,500",
+                            "type":"long"
+                    },
+                    {
+                            "random":"1,127",
+                            "type":"long"
+                    },
+                    {
+                            "type":"string",
+                            "value":"username"
+                    }
+            ],
+            "sliceRecordCount":100
+    },
+    "name":"streamreader"
+},
+"writer":{
+    "parameter":{
+            "password":"*****",
+            "batchSize":1024,
+            "connection":[
+                    {
+                            "database":"example_db",
+                            "endpoint":"http://127.0.0.1:8030/",
+                            "table":"table1"
+                    }
+            ],
+            "username":"test"
+    },
+    "name":"doriswriter"
+}
+}
+],
+"setting":{
+"speed":{
+"channel":2
+}
+}
+}
+
+2021-02-23 15:22:57.886 [main] INFO  PerfTrace - PerfTrace traceId=job_-1, isEnable=false, priority=0
+2021-02-23 15:22:57.886 [main] INFO  JobContainer - DataX jobContainer starts job.
+2021-02-23 15:22:57.920 [job-0] INFO  JobContainer - Scheduler starts [1] taskGroups.
+2021-02-23 15:22:57.928 [taskGroup-0] INFO  TaskGroupContainer - taskGroupId=[0] start [2] channels for [2] tasks.
+2021-02-23 15:22:57.935 [taskGroup-0] INFO  Channel - Channel set byte_speed_limit to -1, No bps activated.
+2021-02-23 15:22:57.936 [taskGroup-0] INFO  Channel - Channel set record_speed_limit to -1, No tps activated.
+2021-02-23 15:22:57.970 [0-0-1-writer] INFO  DorisWriterTask - connect DorisDB with http://127.0.0.1:8030//api/example_db/table1/_stream_load
+2021-02-23 15:22:57.970 [0-0-0-writer] INFO  DorisWriterTask - connect DorisDB with http://127.0.0.1:8030//api/example_db/table1/_stream_load
+
+2021-02-23 15:23:00.941 [job-0] INFO  JobContainer - PerfTrace not enable!
+2021-02-23 15:23:00.946 [job-0] INFO  JobContainer -
+任务启动时刻                    : 2021-02-23 15:22:57
+任务结束时刻                    : 2021-02-23 15:23:00
+任务总计耗时                    :                  3s
+任务平均流量                    :            1.56KB/s
+记录写入速度                    :             66rec/s
+读出记录总数                    :                 200
+读写失败总数                    :                   0
+```
+
+## 参数说明
+
+| 配置项          | 是否必须 | 类型  | 默认值 |         描述   |
+| :-------------- | :------: | ------ |------------- |-------|
+| endpoint         |    是    | string | 无     | Doris 的HTTP连接方式，只需要写到主机和端口即可，具体路径插件会自动拼装 ｜
+| username        |    是    | string | 无     | HTTP 签名验证帐号 |
+| password        |    否    | string | 无     | HTTP 签名验证密码 |
+| table           |    是    | string | 无     | 所选取的需要同步的表名|
+| column          |    否    | list | 无     |  所配置的表中需要同步的列名集合，详细描述见[rdbmswriter](rdbmswriter.md) ｜
+| batchSize       |    否    | int | 1024   | 定义了插件和数据库服务器端每次批量数据获取条数，调高该值可能导致 DataX 出现OOM或者目标数据库事务提交失败导致挂起 |
+
+### column
+
+该插件中的 `column` 不是必须项，如果没有配置该项，或者配置为 `["*"]` ， 则按照 reader 插件获取的字段值进行顺序拼装。
+否则可以按照如下方式指定需要插入的字段
+
+```json
+{
+  "column": ["siteid","citycode","username"]
+}
+```
+

--- a/plugin/writer/doriswriter/package.xml
+++ b/plugin/writer/doriswriter/package.xml
@@ -1,0 +1,38 @@
+<assembly
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2"
+        xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 http://maven.apache.org/xsd/assembly-component-1.1.2.xsd">
+    <id>release</id>
+    <formats>
+        <format>dir</format>
+        <format>zip</format>
+    </formats>
+    <includeBaseDirectory>false</includeBaseDirectory>
+    <fileSets>
+        <fileSet>
+            <directory>src/main/resources</directory>
+            <includes>
+                <include>*.json</include>
+            </includes>
+            <outputDirectory>plugin/writer/${project.artifactId}</outputDirectory>
+        </fileSet>
+        <fileSet>
+            <directory>target/</directory>
+            <includes>
+                <include>${project.artifactId}-${project.version}.jar</include>
+            </includes>
+            <outputDirectory>plugin/writer/${project.artifactId}</outputDirectory>
+        </fileSet>
+    </fileSets>
+
+    <dependencySets>
+        <dependencySet>
+            <useProjectArtifact>false</useProjectArtifact>
+            <outputDirectory>plugin/writer/${project.artifactId}/libs</outputDirectory>
+            <scope>runtime</scope>
+            <excludes>
+                <exclude>com.wgzhao.datax:datax-common</exclude>
+            </excludes>
+        </dependencySet>
+    </dependencySets>
+</assembly>

--- a/plugin/writer/doriswriter/pom.xml
+++ b/plugin/writer/doriswriter/pom.xml
@@ -1,0 +1,62 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.wgzhao.datax</groupId>
+        <artifactId>datax-all</artifactId>
+        <version>3.2.2-SNAPSHOT</version>
+        <relativePath>../../../pom.xml</relativePath>
+    </parent>
+    <artifactId>doriswriter</artifactId>
+    <name>doriswriter</name>
+    <packaging>jar</packaging>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.wgzhao.datax</groupId>
+            <artifactId>datax-common</artifactId>
+            <version>${project.version}</version>
+            <exclusions>
+                <exclusion>
+                    <artifactId>slf4j-log4j12</artifactId>
+                    <groupId>org.slf4j</groupId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <dependency>
+            <groupId>com.wgzhao.datax</groupId>
+            <artifactId>plugin-rdbms-util</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>${httpclient.version}</version>
+        </dependency>
+
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-assembly-plugin</artifactId>
+                <configuration>
+                    <descriptors>
+                        <descriptor>package.xml</descriptor>
+                    </descriptors>
+                    <finalName>${project.artifactId}-${project.version}</finalName>
+                </configuration>
+                <executions>
+                    <execution>
+                        <id>release</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>single</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>

--- a/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/DorisWriter.java
+++ b/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/DorisWriter.java
@@ -1,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.wgzhao.datax.plugin.writer.doriswriter;
+
+import com.wgzhao.datax.common.plugin.RecordReceiver;
+import com.wgzhao.datax.common.spi.Writer;
+import com.wgzhao.datax.common.util.Configuration;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DorisWriter
+        extends Writer
+{
+    public static class Job
+            extends Writer.Job
+    {
+
+        private Configuration originalConfig = null;
+
+        @Override
+        public void init()
+        {
+            this.originalConfig = getPluginJobConf();
+            List<Object> connList = originalConfig.getList(Key.CONNECTION);
+            Configuration conn = Configuration.from(connList.get(0).toString());
+            conn.getNecessaryValue(Key.TABLE, DorisWriterErrorCode.REQUIRED_VALUE);
+            conn.getNecessaryValue(Key.ENDPOINT, DorisWriterErrorCode.REQUIRED_VALUE);
+            conn.getNecessaryValue(Key.DATABASE, DorisWriterErrorCode.REQUIRED_VALUE);
+        }
+
+        @Override
+        public void prepare()
+        {
+            //
+        }
+
+        @Override
+        public List<Configuration> split(int mandatoryNumber)
+        {
+            List<Configuration> splitResultConfigs = new ArrayList<>();
+            for (int j = 0; j < mandatoryNumber; j++) {
+                splitResultConfigs.add(this.originalConfig.clone());
+            }
+            return splitResultConfigs;
+        }
+
+        @Override
+        public void post()
+        {
+            //
+        }
+
+        @Override
+        public void destroy()
+        {
+            //
+        }
+    }
+
+    public static class Task
+            extends Writer.Task
+    {
+        private DorisWriterTask DorisWriterTask;
+
+        @Override
+        public void init()
+        {
+            Configuration writerSliceConfig = getPluginJobConf();
+            this.DorisWriterTask = new DorisWriterTask(writerSliceConfig);
+            this.DorisWriterTask.init();
+        }
+
+        @Override
+        public void prepare()
+        {
+            //
+        }
+
+        public void startWrite(RecordReceiver recordReceiver)
+        {
+            this.DorisWriterTask.startWrite(recordReceiver, getTaskPluginCollector());
+        }
+
+        @Override
+        public void post()
+        {
+            //
+        }
+
+        @Override
+        public void destroy()
+        {
+            //
+        }
+    }
+}

--- a/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/DorisWriterErrorCode.java
+++ b/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/DorisWriterErrorCode.java
@@ -1,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.wgzhao.datax.plugin.writer.doriswriter;
+
+import com.wgzhao.datax.common.spi.ErrorCode;
+
+public enum DorisWriterErrorCode
+        implements ErrorCode
+{
+    REQUIRED_VALUE("DorisWriter-00", "缺失必要的值"),
+    ILLEGAL_VALUE("DorisWriter-01", "值非法"),
+    CONF_ERROR("DorisWriter-02", "您的配置错误."),
+    CONNECT_ERROR("DorisWriter-03", "连接错误"),
+    WRITER_ERROR("DorisWriter-04", "写入错误");
+
+    private final String code;
+    private final String description;
+
+    DorisWriterErrorCode(String code, String description)
+    {
+        this.code = code;
+        this.description = description;
+    }
+
+    @Override
+    public String getCode()
+    {
+        return this.code;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return this.description;
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("Code:[%s], Description:[%s]. ", this.code,
+                this.description);
+    }
+}

--- a/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/DorisWriterTask.java
+++ b/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/DorisWriterTask.java
@@ -1,0 +1,173 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.wgzhao.datax.plugin.writer.doriswriter;
+
+import com.wgzhao.datax.common.element.Record;
+import com.wgzhao.datax.common.exception.DataXException;
+import com.wgzhao.datax.common.plugin.RecordReceiver;
+import com.wgzhao.datax.common.plugin.TaskPluginCollector;
+import com.wgzhao.datax.common.util.Configuration;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.http.HttpHeaders;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.DefaultRedirectStrategy;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.UUID;
+
+public class DorisWriterTask
+{
+    private static final Logger LOG = LoggerFactory.getLogger(DorisWriterTask.class);
+    private final Configuration configuration;
+    private String username;
+    private String password;
+    private String loadUrl;
+    private List<String> column;
+    private static final String SEPARATOR = "|";
+    private int batchSize;
+    private HttpClientBuilder httpClientBuilder;
+
+    public DorisWriterTask(Configuration configuration) {this.configuration = configuration;}
+
+    public void init()
+    {
+        List<Object> connList = configuration.getList(Key.CONNECTION);
+        Configuration conn = Configuration.from(connList.get(0).toString());
+        String endpoint = conn.getString(Key.ENDPOINT);
+        String table = conn.getString(Key.TABLE);
+        String database = conn.getString(Key.DATABASE);
+        this.column = configuration.getList(Key.COLUMN, String.class);
+        // 如果 column 填写的是 * ，直接设置为null，方便后续判断
+        if (this.column!= null && this.column.size() == 1 && "*".equals(this.column.get(0))) {
+            this.column = null;
+        }
+        this.batchSize = configuration.getInt(Key.BATCH_SIZE, 1024);
+        this.username = configuration.getString(Key.USERNAME);
+        this.password = configuration.getString(Key.PASSWORD, null);
+        this.loadUrl = String.format("%s/api/%s/%s/_stream_load", endpoint, database, table);
+        this.httpClientBuilder = HttpClients
+                .custom()
+                .setRedirectStrategy(new DefaultRedirectStrategy()
+                {
+                    @Override
+                    protected boolean isRedirectable(String method)
+                    {
+                        return true;
+                    }
+                });
+        LOG.info("connect DorisDB with {}", this.loadUrl);
+    }
+
+    private String basicAuthHeader(String username, String password)
+    {
+        String tokenEncode = username + ":" + password;
+        byte[] encoded = Base64.encodeBase64(tokenEncode.getBytes(StandardCharsets.UTF_8));
+        return "Basic " + new String(encoded);
+    }
+
+    public void startWrite(RecordReceiver recordReceiver, TaskPluginCollector taskPluginCollector)
+    {
+        Record record;
+        int currSize = 0;
+        StringBuilder stringBuilder = new StringBuilder();
+        while ((record = recordReceiver.getFromReader()) != null) {
+            int len = record.getColumnNumber();
+            if (this.column != null && len != this.column.size()) {
+                throw DataXException.asDataXException(
+                        DorisWriterErrorCode.ILLEGAL_VALUE,
+                        String.format("源字段数和目标字段数不匹配，源字段数为%s, 目标字段数为%s", len, this.column.size())
+                );
+            }
+            StringBuilder oneRow = new StringBuilder();
+            for (int i = 0; i < len; i++) {
+                if (record.getColumn(i).getRawData() != null) {
+                    oneRow.append(record.getColumn(i).asString());
+                }
+                if (i < len - 1) {
+                    oneRow.append(SEPARATOR);
+                }
+            }
+            oneRow.append("\n");
+            stringBuilder.append(oneRow);
+            currSize++;
+            if (currSize >= this.batchSize) {
+                stringBuilder.deleteCharAt(stringBuilder.length() - 1);
+                sendData(stringBuilder.toString());
+                currSize = 0;
+                stringBuilder.setLength(0);
+            }
+        }
+        if (stringBuilder.length() > 0) {
+            sendData(stringBuilder.toString());
+        }
+    }
+
+    private void sendData(String content)
+    {
+        try (CloseableHttpClient client = this.httpClientBuilder.build()) {
+            HttpPut put = new HttpPut(this.loadUrl);
+            StringEntity entity = new StringEntity(content, "UTF-8");
+            put.setHeader(HttpHeaders.EXPECT, "100-continue");
+            if (this.username != null && this.password != null) {
+                put.setHeader(HttpHeaders.AUTHORIZATION, basicAuthHeader(this.username, this.password));
+            }
+            put.setHeader("column_separator", SEPARATOR);
+            // the label header is optional, not necessary
+            // use label header can ensure at most once semantics
+            put.setHeader("label", UUID.randomUUID().toString());
+            if (this.column != null ) {
+                put.setHeader("columns", String.join(",", this.column));
+            }
+            put.setEntity(entity);
+
+            try (CloseableHttpResponse response = client.execute(put)) {
+                String loadResult = "";
+                if (response.getEntity() != null) {
+                    loadResult = EntityUtils.toString(response.getEntity());
+                }
+                int statusCode = response.getStatusLine().getStatusCode();
+                // statusCode 200 just indicates that doris be service is ok, not stream load
+                // you should see the output content to find whether stream load is success
+                if (statusCode != 200) {
+                    throw DataXException.asDataXException(
+                            DorisWriterErrorCode.WRITER_ERROR,
+                            String.format("Stream load failed, statusCode=%s load result=%s", statusCode, loadResult)
+                    );
+                }
+            }
+        }
+        catch (IOException e) {
+            throw DataXException.asDataXException(
+                    DorisWriterErrorCode.CONNECT_ERROR,
+                    String.format("Failed to connect Doris server with: %s, %s", this.loadUrl, e)
+            );
+        }
+    }
+}

--- a/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/Key.java
+++ b/plugin/writer/doriswriter/src/main/java/com/wgzhao/datax/plugin/writer/doriswriter/Key.java
@@ -1,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.wgzhao.datax.plugin.writer.doriswriter;
+
+public class Key
+{
+    public static final String ENDPOINT = "endpoint";
+    public static final String COLUMN = "column";
+    public static final String USERNAME = "username";
+    public static final String PASSWORD = "password";
+    public static final String DATABASE = "database";
+    public static final String TABLE = "table";
+    public static final String CONNECTION = "connection";
+    public static final String BATCH_SIZE = "batchSize";
+}

--- a/plugin/writer/doriswriter/src/main/resources/plugin.json
+++ b/plugin/writer/doriswriter/src/main/resources/plugin.json
@@ -1,0 +1,6 @@
+{
+  "name": "doriswriter",
+  "class": "com.wgzhao.datax.plugin.writer.doriswriter.DorisWriter",
+  "description": "writer data into DorisDB",
+  "developer": "wgzhao"
+}

--- a/plugin/writer/doriswriter/src/main/resources/plugin_job_template.json
+++ b/plugin/writer/doriswriter/src/main/resources/plugin_job_template.json
@@ -1,0 +1,17 @@
+{
+  "writer": {
+    "name": "doriswriter",
+    "parameter": {
+      "username": "test",
+      "password": "123456",
+      "batchSize": 1024,
+      "connection": [
+        {
+          "table": "table1",
+          "database": "example_db",
+          "endpoint": "http://127.0.0.1:8030/"
+        }
+      ]
+    }
+  }
+}

--- a/pom.xml
+++ b/pom.xml
@@ -119,6 +119,7 @@
         <module>plugin/writer/cassandrawriter</module>
         <module>plugin/writer/clickhousewriter</module>
         <module>plugin/writer/dbffilewriter</module>
+        <module>plugin/writer/doriswriter</module>
         <module>plugin/writer/elasticsearchwriter</module>
         <module>plugin/writer/ftpwriter</module>
         <module>plugin/writer/greenplumwriter</module>

